File: src/main/java/com/warren/bt/RetryUntilSuccess.java
package com.warren.bt;

public class RetryUntilSuccess extends Decorator {
  private final int maxAttempts;
  private int currentAttempts = 0;

  public RetryUntilSuccess(Node child, int maxAttempts) {
    super(child);
    this.maxAttempts = maxAttempts;
  }

  public RetryUntilSuccess(Node child) {
    this(child, -1);
  }

  @Override
  public Status tick() {
    if (child == null) {
      return Status.FAILURE;
    }

    Status status = child.tick();

    if (status == Status.SUCCESS) {
      reset();
      return Status.SUCCESS;
    }

    if (status == Status.RUNNING) {
      return Status.RUNNING;
    }

    // Child failed
    currentAttempts++;

    // Check if we've exceeded max attempts
    if (maxAttempts > 0 && currentAttempts >= maxAttempts) {
      reset();
      return Status.FAILURE;
    }

    // Reset child for retry
    child.reset();
    return Status.RUNNING;
  }

  @Override
  public void reset() {
    super.reset();
    currentAttempts = 0;
  }
}

File: src/main/kotlin/com/warren/bt/RetryUntilSuccess.kt
package com.warren.bt

class RetryUntilSuccess(private val maxAttempts: Int = -1) : Decorator("RetryUntilSuccess") {
    private var attempts = 0

    override fun tick(): Status {
        val c = child ?: return Status.FAILURE
        val s = c.tick()
        return when (s) {
            Status.SUCCESS -> Status.SUCCESS
            Status.RUNNING -> Status.RUNNING
            Status.FAILURE -> {
                attempts++
                if (maxAttempts >= 0 && attempts >= maxAttempts) {
                    Status.FAILURE
                } else {
                    c.reset()
                    Status.RUNNING
                }
            }
        }
    }

    override fun onReset() {
        attempts = 0
    }
}
